pkg/antctl/raw/supportbundle/command.go

```diff
@@ -692,9 +702,165 @@ func processResults(resultMap map[string]error, dir string) error {
 		err = writeFailedNodes(dir, failedNodes)
 	}
 
+	// download logs from kubernetes api
+	if failedNodes != nil {
+		err = downloadFallbackAgentBundleFromKubernetes(ctx, antreaClientset, k8sClient, failedNodes, dir)
+		if err != nil {
+			fmt.Println("Failed to download agent bundle from kubernetes api: " + err.Error())
+		} else {
+			allFailed = false
+		}
+	}
+	if controllerFailed {
+		err = downloadFallbackControllerBundleFromKubernetes(ctx, antreaClientset, k8sClient, dir)
+		if err != nil {
+			fmt.Println("Failed to download controller bundle from kubernetes api: " + err.Error())
+		} else {
+			allFailed = false
+		}
+	}
+
 	if allFailed {
 		return fmt.Errorf("no data was collected: %s", genErrorMsg(resultMap))
 	} else {
 		return err
 	}
 }
+
+func downloadFallbackControllerBundleFromKubernetes(ctx context.Context, antreaClientset antrea.Interface, k8sClient kubernetes.Interface, dir string) error {
+	var errors []error
+	tmpDir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
+	if err != nil {
+		return err
+	}
+	defer defaultFS.RemoveAll(tmpDir)
+
+	var podRef *corev1.ObjectReference
+	if err := func() error {
+		controllerInfo, err := antreaClientset.CrdV1beta1().AntreaControllerInfos().Get(ctx, v1beta1.AntreaControllerInfoResourceName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		podRef = &controllerInfo.PodRef
+		data, err := yaml.Marshal(controllerInfo)
+		if err != nil {
+			return err
+		}
+		if err := afero.WriteFile(defaultFS, filepath.Join(dir, "controllerinfo"), data, 0644); err != nil {
+			return err
+		}
+		return nil
+	}(); err != nil {
+		errors = append(errors, err)
+	}
+	errors = append(errors, err)
+	if podRef != nil {
+		pod, err := k8sClient.CoreV1().Pods(podRef.Namespace).Get(ctx, podRef.Name, metav1.GetOptions{})
+		if err == nil {
+			err = downloadAndPackPodLogs(ctx, k8sClient, pod, dir, tmpDir)
+			errors = append(errors, err)
+		}
+	}
+	return utilerror.NewAggregate(errors)
```

Comment:
By the way, I am actually not sure we need an aggregate for this function, since there is only one controller Pod?

Reply:
in the controller case it may seems a bit of complicated. I think the use of aggregated error is to get any many information as possible in a `broken` environment, not only for multiple pods cases.  if we can get the controllerInfo, but failed to marshal and write it to the disk(in very rare cases, maybe can just return instead of aggregate the errors) , this won't affect we continue to retrieve logs from the pod.